From: =?UTF-8?q?Leonard=20G=C3=B6hrs?= <l.goehrs@pengutronix.de>
Date: Tue, 29 Nov 2022 10:55:56 +0100
Subject: [PATCH] ARM: dts: stm32: lxa-tac: add Gen 2 support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This involves a few things:

Switch from OLED to LCD on both generations
-------------------------------------------

Newer revisions of the hardware are shipped with LCD displays instead of OLEDs.
Existing units with OLEDs are upgraded, so there is no need for backward
compatibility.

The generic panel-mipi-dbi-spi driver uses a firmware file for the display
initialization sequence.

Distinguish between device generations
--------------------------------------

There are currently two generations of the LXA TAC with slightly different
hardware configuration.

Add device trees for both and share common parts via a common dtsi.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>
---
 arch/arm/boot/dts/Makefile                     |   2 +
 arch/arm/boot/dts/stm32mp157c-lxa-tac-gen1.dts | 104 ++++
 arch/arm/boot/dts/stm32mp157c-lxa-tac-gen2.dts | 186 +++++++
 arch/arm/boot/dts/stm32mp157c-lxa-tac.dts      | 729 -------------------------
 arch/arm/boot/dts/stm32mp15xc-lxa-tac.dtsi     | 716 ++++++++++++++++++++++++
 5 files changed, 1008 insertions(+), 729 deletions(-)
 create mode 100644 arch/arm/boot/dts/stm32mp157c-lxa-tac-gen1.dts
 create mode 100644 arch/arm/boot/dts/stm32mp157c-lxa-tac-gen2.dts
 delete mode 100644 arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
 create mode 100644 arch/arm/boot/dts/stm32mp15xc-lxa-tac.dtsi

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index efe4152e5846..e489daedacc5 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1252,6 +1252,8 @@ dtb-$(CONFIG_ARCH_STM32) += \
 	stm32mp157c-ev1.dtb \
 	stm32mp157c-ev1-scmi.dtb \
 	stm32mp157c-lxa-mc1.dtb \
+	stm32mp157c-lxa-tac-gen1.dtb \
+	stm32mp157c-lxa-tac-gen2.dtb \
 	stm32mp157c-odyssey.dtb
 dtb-$(CONFIG_MACH_SUN4I) += \
 	sun4i-a10-a1000.dtb \
diff --git a/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen1.dts b/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen1.dts
new file mode 100644
index 000000000000..1fcfdb99022f
--- /dev/null
+++ b/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen1.dts
@@ -0,0 +1,104 @@
+/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause) */
+/*
+ * Copyright (C) 2020 STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021 Rouven Czerwinski, Pengutronix
+ * Copyright (C) 2023 Leonard Göhrs, Pengutronix
+ */
+
+/dts-v1/;
+
+#include "stm32mp157.dtsi"
+#include "stm32mp15xc-lxa-tac.dtsi"
+
+/ {
+	model = "Linux Automation Test Automation Controller (TAC) Gen 1";
+	compatible = "lxa,stm32mp157c-tac-gen1", "oct,stm32mp15xx-osd32", "st,stm32mp157";
+
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&backlight_pwm 1 1000000 0>;
+		brightness-levels = <0 31 63 95 127 159 191 223 255>;
+		default-brightness-level = <7>;
+		power-supply = <&v3v3>;
+	};
+
+	reg_iobus_12v: regulator_iobus_12v {
+		compatible = "regulator-fixed";
+		regulator-name = "12V_IOBUS";
+		regulator-min-microvolt = <12000000>;
+		regulator-max-microvolt = <12000000>;
+		vin-supply = <&reg_12v>;
+		gpio = <&gpioh 13 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+};
+
+&timers1 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+
+	backlight_pwm: pwm {
+		pinctrl-names = "default", "sleep";
+		pinctrl-0 = <&pwm1_pins_c>;
+		pinctrl-1 = <&pwm1_sleep_pins_c>;
+		status = "okay";
+	};
+};
+
+&gpioa {
+	gpio-line-names = "", "", "STACK_CS2", "", \
+	"STACK_CS3", "ETH_GPIO1", "ETH_INT", "", \
+	"", "", "", "", \
+	"", "BOOTROM_LED", "ETH_LAB_LEDRP", "";
+};
+
+&gpioc {
+	gpio-line-names = "", "STACK_CS1", "", "", "", "", "", "", \
+	"", "", "", "";
+};
+
+&i2c1 {
+	powerboard_gpio: pca9570@24 {
+		compatible = "nxp,pca9570";
+		reg = <0x24>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		gpio-line-names = "DUT_PWR_EN", "DUT_PWR_DISCH", \
+		"DUT_PWR_ADCRST", "";
+		status = "okay";
+	};
+};
+
+&spi2 {
+	lmp92064@0 {
+		compatible = "ti,lmp92064";
+		reg = <0>;
+		spi-max-frequency = <5000000>;
+		vdd-supply = <&reg_pb_3v3>;
+		vdig-supply = <&reg_pb_3v3>;
+		reset-gpios = <&powerboard_gpio 2 GPIO_ACTIVE_HIGH>;
+		shunt-resistor-micro-ohms = <15000>;
+	};
+};
+
+&gpu {
+	status = "disabled";
+};
+
+&pinctrl {
+	pwm1_pins_c: pwm1-2 {
+		pins {
+			pinmux = <STM32_PINMUX('E', 11, AF1)>; /* TIM1_CH2 */
+			drive-push-pull;
+			slew-rate = <0>;
+		};
+	};
+
+	pwm1_sleep_pins_c: pwm1-sleep-2 {
+		pins {
+			pinmux = <STM32_PINMUX('E', 11, ANALOG)>; /* TIM1_CH2 */
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen2.dts b/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen2.dts
new file mode 100644
index 000000000000..8fdea9bdd282
--- /dev/null
+++ b/arch/arm/boot/dts/stm32mp157c-lxa-tac-gen2.dts
@@ -0,0 +1,186 @@
+/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause) */
+/*
+ * Copyright (C) 2020 STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021 Rouven Czerwinski, Pengutronix
+ * Copyright (C) 2023 Leonard Göhrs, Pengutronix
+ */
+
+/dts-v1/;
+
+#include "stm32mp157.dtsi"
+#include "stm32mp15xc-lxa-tac.dtsi"
+
+/ {
+	model = "Linux Automation Test Automation Controller (TAC) Gen 2";
+	compatible = "lxa,stm32mp157c-tac-gen2", "oct,stm32mp15xx-osd32", "st,stm32mp157";
+
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&led_pwm 3 1000000 0>;
+		brightness-levels = <0 31 63 95 127 159 191 223 255>;
+		default-brightness-level = <7>;
+		power-supply = <&v3v3>;
+	};
+
+	reg_iobus_12v: regulator_iobus_12v {
+		compatible = "regulator-fixed";
+		regulator-name = "12V_IOBUS";
+		regulator-min-microvolt = <12000000>;
+		regulator-max-microvolt = <12000000>;
+		vin-supply = <&reg_12v>;
+		gpio = <&gpioh 13 GPIO_ACTIVE_LOW>;
+	};
+
+	led-controller-1 {
+		compatible = "pwm-leds-multicolor";
+
+		multi-led {
+			color = <LED_COLOR_ID_RGB>;
+			function = LED_FUNCTION_STATUS;
+			max-brightness = <65535>;
+
+			led-red {
+				active-low;
+				pwms = <&led_pwm 0 1000000 0>;
+				color = <LED_COLOR_ID_RED>;
+			};
+
+			led-green {
+				active-low;
+				pwms = <&led_pwm 2 1000000 0>;
+				color = <LED_COLOR_ID_GREEN>;
+			};
+
+			led-blue {
+				active-low;
+				pwms = <&led_pwm 1 1000000 0>;
+				color = <LED_COLOR_ID_BLUE>;
+			};
+		};
+	};
+
+	led-controller-2 {
+		compatible = "gpio-leds";
+
+		led-5 {
+			label = "tac:green:iobus";
+			gpios = <&gpiog 1 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-6 {
+			label = "tac:green:can";
+			gpios = <&gpiof 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-7 {
+			label = "tac:green:out0";
+			gpios = <&gpiob 8 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-8 {
+			label = "tac:green:out1";
+			gpios = <&gpiog 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-9 {
+			label = "tac:green:uarttx";
+			gpios = <&gpiod 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-10 {
+			label = "tac:green:uartrx";
+			gpios = <&gpiof 6 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-11 {
+			label = "tac:green:usbh1";
+			gpios = <&gpioc 8 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-12 {
+			label = "tac:green:usbh2";
+			gpios = <&gpiod 6 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-13 {
+			label = "tac:green:usbh3";
+			gpios = <&gpiob 9 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-14 {
+			label = "tac:green:usbg";
+			gpios = <&gpiod 14 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "usb-gadget";
+		};
+
+		led-15 {
+			label = "tac:green:dutpwr";
+			gpios = <&gpioa 15 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&timers8 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+
+	led_pwm: pwm {
+		pinctrl-names = "default", "sleep";
+		pinctrl-0 = <&pwm8_pins_b>;
+		pinctrl-1 = <&pwm8_sleep_pins_b>;
+		status = "okay";
+	};
+};
+
+&gpioa {
+	gpio-line-names = "", "", "DUT_PWR_EN", "", \
+	"STACK_CS3", "ETH_GPIO1", "ETH_INT", "", \
+	"", "", "", "", \
+	"", "BOOTROM_LED", "ETH_LAB_LEDRP", "";
+};
+
+&gpioc {
+	gpio-line-names = "", "DUT_PWR_DISCH", "", "", "", "", "", "", \
+	"", "", "", "";
+};
+
+&spi2 {
+	lmp92064@0 {
+		compatible = "ti,lmp92064";
+		reg = <0>;
+		spi-max-frequency = <5000000>;
+		vdd-supply = <&reg_pb_3v3>;
+		vdig-supply = <&reg_pb_3v3>;
+		reset-gpios = <&gpioa 4 GPIO_ACTIVE_HIGH>;
+		shunt-resistor-micro-ohms = <15000>;
+	};
+};
+
+&gpu {
+	status = "disabled";
+};
+
+&pinctrl {
+	pwm8_pins_b: pwm8-1 {
+		pins {
+			pinmux = <STM32_PINMUX('I', 5, AF3)>, /* TIM8_CH1 */
+				 <STM32_PINMUX('I', 6, AF3)>, /* TIM8_CH2 */
+				 <STM32_PINMUX('I', 7, AF3)>, /* TIM8_CH3 */
+				 <STM32_PINMUX('C', 9, AF3)>; /* TIM8_CH4 */
+			drive-push-pull;
+			slew-rate = <0>;
+		};
+	};
+
+	pwm8_sleep_pins_b: pwm8-sleep-1 {
+		pins {
+			pinmux = <STM32_PINMUX('I', 5, ANALOG)>, /* TIM8_CH1 */
+				 <STM32_PINMUX('I', 6, ANALOG)>, /* TIM8_CH2 */
+				 <STM32_PINMUX('I', 7, ANALOG)>, /* TIM8_CH3 */
+				 <STM32_PINMUX('C', 9, ANALOG)>; /* TIM8_CH4 */
+		};
+	};
+};
+
diff --git a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts b/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
deleted file mode 100644
index 13bd0068d517..000000000000
--- a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
+++ /dev/null
@@ -1,729 +0,0 @@
-/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause) */
-/*
- * Copyright (C) 2020 STMicroelectronics - All Rights Reserved
- * Copyright (C) 2021 Rouven Czerwinski, Pengutronix
- */
-
-/dts-v1/;
-
-#include "stm32mp157.dtsi"
-#include "stm32mp15xc.dtsi"
-#include "stm32mp15xx-osd32.dtsi"
-#include "stm32mp15xxac-pinctrl.dtsi"
-
-#include <dt-bindings/gpio/gpio.h>
-#include <dt-bindings/input/input.h>
-#include <dt-bindings/pwm/pwm.h>
-
-/ {
-	model = "Linux Automation Test Automation Controller (TAC)";
-	compatible = "lxa,stm32mp157c-tac", "oct,stm32mp15xx-osd32", "st,stm32mp157";
-
-	aliases {
-		ethernet0 = &ethernet0;
-		ethernet1 = &port_uplink;
-		ethernet2 = &port_dut;
-		mmc1 = &sdmmc2;
-		serial0 = &uart4;
-		serial1 = &usart3;
-	};
-
-	chosen {
-		stdout-path = &uart4;
-	};
-
-	led-controller-0 {
-		compatible = "gpio-leds";
-
-		led-0 {
-			label = "tac:green:user1";
-			gpios = <&gpiof 10 GPIO_ACTIVE_HIGH>;
-			linux,default-trigger = "heartbeat";
-		};
-
-		led-1 {
-			label = "tac:green:user2";
-			gpios = <&gpiog 7 GPIO_ACTIVE_HIGH>;
-		};
-
-		led-2 {
-			label = "tac:green:statusdut";
-			gpios = <&gpioa 13 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "default-on";
-		};
-
-		/* led-3 and led-4 are internally connected antiparallel to one
-		 * another inside the ethernet jack like this:
-		 * GPIOA14 ---+---|led-3|>--+--- GPIOD15
-		 *            +--<|led-4|---+
-		 * E.g. only one of the LEDs can be illuminated at a time while
-		 * the other output must be driven low.
-		 * This should likely be implemented using a multi color LED
-		 * driver for antiparallel LEDs. */
-		led-3 {
-			label = "tac:green:statuslab";
-			gpios = <&gpioa 14 GPIO_ACTIVE_HIGH>;
-			linux,default-trigger = "default-on";
-		};
-
-		led-4 {
-			label = "tac:orange:statuslab";
-			gpios = <&gpiod 15 GPIO_ACTIVE_HIGH>;
-		};
-	};
-
-	gpio-keys {
-		compatible = "gpio-keys";
-
-		user-button1 {
-			label = "USER_BTN";
-			linux,code = <KEY_HOME>;
-			gpios = <&gpioi 11 GPIO_ACTIVE_LOW>;
-		};
-
-		user-button2 {
-			label = "USER_BTN2";
-			linux,code = <KEY_ESC>;
-			gpios = <&gpioe 7 GPIO_ACTIVE_LOW>;
-		};
-	};
-
-	/* supplied by either barrel connector or PoE */
-	reg_12v: regulator_12v {
-		compatible = "regulator-fixed";
-		regulator-name = "12V";
-		regulator-min-microvolt = <12000000>;
-		regulator-max-microvolt = <12000000>;
-		regulator-always-on;
-	};
-
-	reg_5v: regulator_5v {
-		compatible = "regulator-fixed";
-		regulator-name = "5V";
-		regulator-min-microvolt = <5000000>;
-		regulator-max-microvolt = <5000000>;
-		regulator-always-on;
-		vin-supply = <&reg_12v>;
-	};
-
-	reg_1v2: regulator_1v2 {
-		compatible = "regulator-fixed";
-		regulator-name = "1V2";
-		regulator-min-microvolt = <1200000>;
-		regulator-max-microvolt = <1200000>;
-		regulator-always-on;
-		vin-supply = <&reg_5v>;
-	};
-
-	reg_pb_5v: regulator_pb_5v {
-		compatible = "regulator-fixed";
-		regulator-name = "5V_POWERBOARD";
-		regulator-min-microvolt = <5000000>;
-		regulator-max-microvolt = <5000000>;
-		regulator-always-on;
-		vin-supply = <&reg_5v>;
-	};
-
-	reg_pb_3v3: regulator_pb_3v3 {
-		compatible = "regulator-fixed";
-		regulator-name = "3V3_POWERBOARD";
-		regulator-min-microvolt = <3300000>;
-		regulator-max-microvolt = <3300000>;
-		regulator-always-on;
-		vin-supply = <&reg_pb_5v>;
-	};
-
-	iobus_curr: current-sense-amplifier-0 {
-		compatible = "current-sense-amplifier";
-		io-channels = <&adc1 5>;
-		#io-channel-cells = <0>;
-		label = "iobus-curr";
-		sense-resistor-micro-ohms = <2200000000>; /* 2.2k */
-		sense-gain-mult = <653>;
-		sense-gain-div = <1000000>;
-	};
-
-	iobus_volt: voltage-divider {
-		compatible = "voltage-divider";
-		io-channels = <&adc1 9>;
-		#io-channel-cells = <0>;
-		label = "iobus-volt";
-		output-ohms = <1000>;
-		full-ohms = <6100>;
-	};
-
-	iobus {
-		compatible = "iio-hwmon";
-		io-channels = <&iobus_volt>, <&iobus_curr>;
-	};
-
-	host_curr: current-sense-amplifier-1 {
-		compatible = "current-sense-amplifier";
-		io-channels = <&adc1 13>;
-		#io-channel-cells = <0>;
-		label = "host-curr";
-		sense-resistor-micro-ohms = <1300000000>; /* 1.3k */
-		sense-gain-mult = <653>;
-		sense-gain-div = <1000000>;
-	};
-
-	host_1_curr: current-sense-amplifier-2 {
-		compatible = "current-sense-amplifier";
-		io-channels = <&adc1 15>;
-		#io-channel-cells = <0>;
-		label = "host-1-curr";
-		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
-		sense-gain-mult = <653>;
-		sense-gain-div = <1000000>;
-	};
-
-	host_2_curr: current-sense-amplifier-3 {
-		compatible = "current-sense-amplifier";
-		io-channels = <&adc1 0>; /* via ANA0 */
-		#io-channel-cells = <0>;
-		label = "host-2-curr";
-		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
-		sense-gain-mult = <653>;
-		sense-gain-div = <1000000>;
-	};
-
-	host_3_curr: current-sense-amplifier-4 {
-		compatible = "current-sense-amplifier";
-		io-channels = <&adc1 1>; /* via ANA1 */
-		#io-channel-cells = <0>;
-		label = "host-3-curr";
-		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
-		sense-gain-mult = <653>;
-		sense-gain-div = <1000000>;
-	};
-
-	usb {
-		compatible = "iio-hwmon";
-		io-channels = <&host_curr>, <&host_1_curr>, <&host_2_curr>, <&host_3_curr>;
-	};
-};
-
-&pwr_regulators {
-	vdd-supply = <&vdd>;
-	vdd_3v3_usbfs-supply = <&vdd_usb>;
-};
-
-&hash1 {
-	status = "okay";
-};
-
-&crc1 {
-	status = "okay";
-};
-
-&cryp1 {
-	status = "okay";
-};
-
-&vrefbuf {
-	regulator-min-microvolt = <2500000>;
-	regulator-max-microvolt = <2500000>;
-	vdda-supply = <&vdda>;
-	status = "okay";
-};
-
-&adc {
-	pinctrl-names = "default";
-	pinctrl-0 = <&adc1_ain_pins_a>;
-	vdd-supply = <&vdd>;
-	vdda-supply = <&vdda>;
-	vref-supply = <&vrefbuf>;
-	status = "okay";
-
-	adc1: adc@0 {
-		st,adc-channels = <0 1 2 5 6 9 10 11 13 14 15>;
-		st,min-sample-time-nsecs = <5000>;
-		#address-cells = <1>;
-		#size-cells = <0>;
-		status = "okay";
-		channel@0 {
-			reg = <0>;
-			label = "HOST_2_CURR_FB";
-		};
-		channel@1 {
-			reg = <1>;
-			label = "HOST_3_CURR_FB";
-		};
-		channel@2 {
-			reg = <2>;
-			label = "OUT_0_FB";
-		};
-		channel@5 {
-			reg = <5>;
-			label = "IOBUS_CURR_FB";
-		};
-		channel@9 {
-			reg = <9>;
-			label = "IOBUS_VOLT_FB";
-		};
-		channel@10 {
-			reg = <10>;
-			label = "OUT_1_FB";
-		};
-		channel@13 {
-			reg = <13>;
-			label = "HOST_CURR_FB";
-		};
-		channel@15 {
-			reg = <15>;
-			label = "HOST_1_CURR_FB";
-		};
-	};
-
-	adc2: adc@100 {
-		st,adc-channels = <12>;
-		st,min-sample-time-nsecs = <500000>;
-		#address-cells = <1>;
-		#size-cells = <0>;
-		status = "okay";
-		channel@12 {
-			reg = <12>;
-			label = "TEMP_INTERNAL";
-		};
-	};
-};
-
-&rtc {
-	status = "okay";
-};
-
-&iwdg2 {
-	timeout-sec = <8>;
-	status = "okay";
-};
-
-&timers2 {
-	/* spare dmas for other usage */
-	/delete-property/dmas;
-	/delete-property/dma-names;
-	status = "okay";
-	timer@1 {
-		status = "okay";
-	};
-};
-
-&timers3 {
-	/* spare dmas for other usage */
-	/delete-property/dmas;
-	/delete-property/dma-names;
-	status = "okay";
-	timer@2 {
-		status = "okay";
-	};
-};
-
-&timers4 {
-	/* spare dmas for other usage */
-	/delete-property/dmas;
-	/delete-property/dma-names;
-	status = "okay";
-	timer@3 {
-		status = "okay";
-	};
-};
-
-&dts {
-	status = "okay";
-};
-
-&gpioa {
-	gpio-line-names = "", "", "STACK_CS2", "", \
-	"STACK_CS3", "ETH_GPIO1", "ETH_INT", "", \
-	"", "", "", "", \
-	"", "BOOTROM_LED", "ETH_LAB_LEDRP", "";
-};
-
-&gpiob {
-	gpio-line-names = "", "", "", "", "", "", "USB_RESET", "", \
-	"", "", "", "", "", "", "", "";
-};
-
-&gpioc {
-	gpio-line-names = "", "STACK_CS1", "", "", "", "", "", "", \
-	"", "", "", "";
-};
-
-&gpiod {
-	gpio-line-names = "", "", "", "", "TP38", "TP39", "", "", \
-	"TP41", "TP42", "OLED_DC", "", "", "ETH_CS", "", "ETH_LAB_LEDRN";
-};
-
-&gpioe {
-	gpio-line-names = "TP35", "", "", "", "CAN_1_120R", "", "", "USER_BTN2", \
-	"TP48", "UART_TX_EN", "UART_RX_EN", "TP24", "", "TP25", "TP26", "TP27";
-};
-
-&gpiof {
-	gpio-line-names = "TP36", "TP37", "", "", "OLED_CS", "", "", "", \
-	"", "", "USER_LED1", "", "STACK_CS0", "", "", "";
-};
-
-&gpiog {
-	gpio-line-names = "ETH_RESET", "", "", "", \
-	"", "IOBUS_FLT_FB", "", "USER_LED2", \
-	"ETH1_PPS_A", "CAN_0_120R", "TP49", "", \
-	"", "", "", "";
-};
-
-&gpioh {
-	gpio-line-names = "", "", "OUT_1", "OUT_0", "OLED_RESET", "", "", "", \
-	"", "", "ETH1_PPS_B", "ETH_GPIO2", "", "IOBUS_PWR_EN", "", "TP33";
-};
-
-&gpioi {
-	gpio-line-names = "TIM_RTS", "", "", "", \
-	"DEVICE_DATA_EN", "", "", "", \
-	"ETH_WOL", "TP43", "", "USER_BTN";
-};
-
-&gpioz {
-	gpio-line-names = "HWID0", "HWID1", "HWID2", "HWID3", \
-	"", "", "HWID4", "HWID5";
-};
-
-&gpu {
-	status = "disabled";
-};
-
-&pmic {
-	regulators {
-		buck1-supply = <&reg_5v>;	/* VIN */
-		buck2-supply = <&reg_5v>;	/* VIN */
-		buck3-supply = <&reg_5v>;	/* VIN */
-		buck4-supply = <&reg_5v>;	/* VIN */
-		ldo2-supply = <&reg_5v>;	/* PMIC_LDO25IN */
-		ldo4-supply = <&reg_5v>;	/* VIN */
-		ldo5-supply = <&reg_5v>;	/* PMIC_LDO25IN */
-		vref_ddr-supply = <&reg_5v>;	/* VIN */
-		boost-supply = <&reg_5v>;	/* PMIC_BSTIN */
-		pwr_sw2-supply = <&bst_out>;	/* PMIC_SWIN */
-	};
-};
-
-&sdmmc2 {
-	pinctrl-names = "default", "opendrain", "sleep";
-	pinctrl-0 = <&sdmmc2_b4_pins_a &sdmmc2_d47_pins_b>;
-	pinctrl-1 = <&sdmmc2_b4_od_pins_a &sdmmc2_d47_pins_b>;
-	pinctrl-2 = <&sdmmc2_b4_sleep_pins_a &sdmmc2_d47_sleep_pins_b>;
-	bus-width = <8>;
-	mmc-ddr-3_3v;
-	no-1-8-v;
-	no-sd;
-	no-sdio;
-	non-removable;
-	st,neg-edge;
-	vmmc-supply = <&v3v3>;
-	status = "okay";
-};
-
-&uart4 {
-	label = "debug";
-	pinctrl-names = "default";
-	pinctrl-0 = <&uart4_pins_a>;
-	/delete-property/dmas;
-	/delete-property/dma-names;
-	status = "okay";
-};
-
-&ethernet0 {
-	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&ethernet0_rgmii_pins_d>;
-	pinctrl-1 = <&ethernet0_rgmii_sleep_pins_d>;
-	phy-mode = "rgmii-id";
-	assigned-clocks = <&rcc ETHCK_K>, <&rcc PLL4_P>;
-	assigned-clock-parents = <&rcc PLL4_P>;
-	assigned-clock-rates = <125000000>; /* Clock PLL4 to 750Mhz in ATF/U-Boot */
-	st,eth-clk-sel;
-	status = "okay";
-
-	fixed-link {
-		speed = <1000>;
-		full-duplex;
-	};
-};
-
-&i2c1 {
-	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&i2c1_pins_b>;
-	pinctrl-1 = <&i2c1_sleep_pins_b>;
-	status = "okay";
-
-	powerboard_gpio: pca9570@24 {
-		compatible = "nxp,pca9570";
-		reg = <0x24>;
-		gpio-controller;
-		#gpio-cells = <2>;
-		gpio-line-names = "IO0", "IO1", "IO2", "IO3";
-		status = "okay";
-	};
-
-	powerboard_eeprom: eeprom@50 {
-		compatible = "atmel,24c02";
-		reg = <0x50>;
-		vcc-supply = <&v3v3>;
-	};
-
-	lm75ad@48 {
-		compatible = "national,lm75a";
-		reg = <0x48>;
-		status = "disabled";
-	};
-};
-
-baseboard_eeprom: &sip_eeprom {
-};
-
-&i2c5 {
-	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&i2c5_pins_b>;
-	pinctrl-1 = <&i2c5_sleep_pins_b>;
-	status = "okay";
-	/delete-property/dmas;
-	/delete-property/dma-names;
-
-	usbhub: usbhub@2c {
-		compatible ="microchip,usb2514b";
-		reg = <0x2c>;
-		reset-gpios = <&gpiob 6 GPIO_ACTIVE_LOW>;
-		vdd-supply = <&v3v3>;
-	};
-};
-
-&spi2 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&spi2_pins_b>;
-	cs-gpios = <&gpiof 12 GPIO_ACTIVE_LOW>;
-	status = "okay";
-
-	lmp92064@0 {
-		compatible = "ti,lmp92064";
-		reg = <0>;
-		spi-max-frequency = <5000000>;
-		vdd-supply = <&reg_pb_3v3>;
-		vdig-supply = <&reg_pb_3v3>;
-		reset-gpios = <&powerboard_gpio 2 GPIO_ACTIVE_HIGH>;
-		shunt-resistor-micro-ohms = <15000>;
-	};
-};
-
-&spi5 {
-	/* spare dmas for other usage */
-	/delete-property/dmas;
-	/delete-property/dma-names;
-	pinctrl-names = "default";
-	pinctrl-0 = <&spi5_pins_a>;
-	cs-gpios = <&gpiod 13 GPIO_ACTIVE_LOW>;
-	status = "okay";
-
-	ksz9563: ksz9563@0 {
-		compatible = "microchip,ksz9563";
-		reg = <0>;
-
-		reset-gpios = <&gpiog 0 GPIO_ACTIVE_LOW>;
-		spi-max-frequency = <44000000>;
-		spi-cpha;
-		spi-cpol;
-
-		interrupt-parent = <&gpioa>;
-		interrupts = <6 IRQ_TYPE_EDGE_RISING>;
-
-		status = "okay";
-
-		ports {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			port_dut: port@0 {
-				reg = <0>;
-				label = "dut";
-			};
-			port_uplink: port@1 {
-				reg = <1>;
-				label = "uplink";
-			};
-			port_cpu: port@2 {
-				reg = <2>;
-				label = "cpu";
-
-				ethernet = <&ethernet0>;
-
-				phy-mode = "rgmii-id";
-				rx-internal-delay-ps = <2000>;
-				tx-internal-delay-ps = <2000>;
-
-				fixed-link {
-					speed = <1000>;
-					full-duplex;
-				};
-			};
-		};
-	};
-};
-
-&usbotg_hs {
-	phys = <&usbphyc_port1 0>;
-	phy-names = "usb2-phy";
-	dr_mode = "peripheral";
-	vusb_d-supply = <&vdd_usb>;
-	vusb_a-supply = <&reg18>;
-	status = "okay";
-};
-
-&usbh_ehci {
-	phys = <&usbphyc_port0>;
-	phy-names = "usb";
-	status = "okay";
-};
-
-&spi4 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&spi4_pins_a>;
-	cs-gpios = <&gpiof 4 GPIO_ACTIVE_LOW>;
-	status = "okay";
-
-	oled: display@0 {
-		compatible = "solomon,ssd1306";
-		reg = <0>;
-		spi-max-frequency = <10000000>;
-		buswidth = <8>;
-		fps = <20>;
-		reset-gpios = <&gpioh 4 GPIO_ACTIVE_LOW>;
-		dc-gpios = <&gpiod 10 GPIO_ACTIVE_HIGH>;
-		debug = <0x0>;
-	};
-};
-
-&usbphyc {
-	status = "okay";
-};
-
-&usbphyc_port0 {
-	phy-supply = <&vdd_usb>;
-};
-
-&usbphyc_port1 {
-	phy-supply = <&vdd_usb>;
-};
-
-&m_can1 {
-	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&m_can1_pins_b>;
-	pinctrl-1 = <&m_can1_sleep_pins_b>;
-	status = "okay";
-};
-
-&m_can2 {
-	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&m_can2_pins_a>;
-	pinctrl-1 = <&m_can2_sleep_pins_a>;
-	status = "okay";
-};
-
-&usart3 {
-	label = "dut";
-	uart-has-rtscts;
-
-	/delete-property/dmas;
-	/delete-property/dma-names;
-
-	pinctrl-names = "default";
-	pinctrl-0 = <&usart3_pins_e>;
-
-	status = "okay";
-};
-
-&pinctrl {
-	adc1_ain_pins_a: adc1-ain-0 {
-		pins {
-			pinmux = <STM32_PINMUX('F', 11, ANALOG)>, /* ADC1_INP2 */
-				 <STM32_PINMUX('B', 1, ANALOG)>, /* ADC1_INP5 */
-				 <STM32_PINMUX('B', 0, ANALOG)>, /* ADC1_INP9 */
-				 <STM32_PINMUX('C', 0, ANALOG)>, /* ADC1_INP10 */
-				 <STM32_PINMUX('C', 3, ANALOG)>, /* ADC1_INP13 */
-				 <STM32_PINMUX('A', 3, ANALOG)>; /* ADC1_INP15 */
-		};
-	};
-	ethernet0_rgmii_pins_d: rgmii-1 {
-		pins1 {
-			pinmux = <STM32_PINMUX('G', 4, AF11)>, /* ETH_RGMII_GTX_CLK */
-				 <STM32_PINMUX('G', 13, AF11)>, /* ETH_RGMII_TXD0 */
-				 <STM32_PINMUX('G', 14, AF11)>, /* ETH_RGMII_TXD1 */
-				 <STM32_PINMUX('C', 2, AF11)>, /* ETH_RGMII_TXD2 */
-				 <STM32_PINMUX('E', 2, AF11)>, /* ETH_RGMII_TXD3 */
-				 <STM32_PINMUX('B', 11, AF11)>; /* ETH_RGMII_TX_CTL */
-			bias-disable;
-			drive-push-pull;
-			slew-rate = <1>;
-		};
-		pins2 {
-			pinmux = <STM32_PINMUX('C', 4, AF11)>, /* ETH_RGMII_RXD0 */
-				 <STM32_PINMUX('C', 5, AF11)>, /* ETH_RGMII_RXD1 */
-				 <STM32_PINMUX('H', 6, AF11)>, /* ETH_RGMII_RXD2 */
-				 <STM32_PINMUX('H', 7, AF11)>, /* ETH_RGMII_RXD3 */
-				 <STM32_PINMUX('A', 1, AF11)>, /* ETH_RGMII_RX_CLK */
-				 <STM32_PINMUX('A', 7, AF11)>; /* ETH_RGMII_RX_CTL */
-			bias-disable;
-		};
-	};
-	ethernet0_rgmii_sleep_pins_d: rgmii-sleep-1 {
-		pins1 {
-			pinmux = <STM32_PINMUX('G', 4, ANALOG)>, /* ETH_RGMII_GTX_CLK */
-				 <STM32_PINMUX('G', 13, ANALOG)>, /* ETH_RGMII_TXD0 */
-				 <STM32_PINMUX('G', 14, ANALOG)>, /* ETH_RGMII_TXD1 */
-				 <STM32_PINMUX('C', 2, ANALOG)>, /* ETH_RGMII_TXD2 */
-				 <STM32_PINMUX('E', 2, ANALOG)>, /* ETH_RGMII_TXD3 */
-				 <STM32_PINMUX('B', 11, ANALOG)>, /* ETH_RGMII_TX_CTL */
-				 <STM32_PINMUX('C', 4, ANALOG)>, /* ETH_RGMII_RXD0 */
-				 <STM32_PINMUX('C', 5, ANALOG)>, /* ETH_RGMII_RXD1 */
-				 <STM32_PINMUX('H', 6, ANALOG)>, /* ETH_RGMII_RXD2 */
-				 <STM32_PINMUX('H', 7, ANALOG)>, /* ETH_RGMII_RXD3 */
-				 <STM32_PINMUX('A', 1, ANALOG)>, /* ETH_RGMII_RX_CLK */
-				 <STM32_PINMUX('A', 7, ANALOG)>; /* ETH_RGMII_RX_CTL */
-		 };
-	};
-	spi2_pins_b: spi2-1 {
-		pins1 {
-			pinmux = <STM32_PINMUX('I', 1, AF5)>, /* SPI2_SCK */
-				 <STM32_PINMUX('I', 3, AF5)>; /* SPI2_MOSI */
-			bias-disable;
-			drive-push-pull;
-		};
-
-		pins2 {
-			pinmux = <STM32_PINMUX('I', 2, AF5)>; /* SPI2_MISO */
-			bias-pull-down;
-		};
-	};
-	spi5_pins_a: spi5-0 {
-		pins1 {
-			pinmux = <STM32_PINMUX('F', 7, AF5)>, /* SPI5_SCK */
-				 <STM32_PINMUX('F', 9, AF5)>; /* SPI5_MOSI */
-			bias-disable;
-			drive-push-pull;
-			slew-rate = <1>;
-		};
-
-		pins2 {
-			pinmux = <STM32_PINMUX('F', 8, AF5)>; /* SPI5_MISO */
-			bias-disable;
-		};
-	};
-	usart3_pins_e: usart3-4 {
-		pins1 {
-			pinmux = <STM32_PINMUX('B', 10, AF7)>, /* USART3_TX */
-				 <STM32_PINMUX('D', 12, AF7)>; /* USART3_RTS */
-			bias-disable;
-			drive-push-pull;
-			slew-rate = <0>;
-		};
-		pins2 {
-			pinmux = <STM32_PINMUX('B', 12, AF8)>, /* USART3_RX */
-				 <STM32_PINMUX('D', 11, AF7)>; /* USART3_CTS_NSS */
-			bias-disable;
-		};
-	};
-};
diff --git a/arch/arm/boot/dts/stm32mp15xc-lxa-tac.dtsi b/arch/arm/boot/dts/stm32mp15xc-lxa-tac.dtsi
new file mode 100644
index 000000000000..8641393e22a6
--- /dev/null
+++ b/arch/arm/boot/dts/stm32mp15xc-lxa-tac.dtsi
@@ -0,0 +1,716 @@
+/* SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause) */
+/*
+ * Copyright (C) 2020 STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021 Rouven Czerwinski, Pengutronix
+ */
+
+#include "stm32mp15xc.dtsi"
+#include "stm32mp15xx-osd32.dtsi"
+#include "stm32mp15xxac-pinctrl.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/pwm/pwm.h>
+
+/ {
+	aliases {
+		ethernet0 = &ethernet0;
+		ethernet1 = &port_uplink;
+		ethernet2 = &port_dut;
+		mmc1 = &sdmmc2;
+		serial0 = &uart4;
+		serial1 = &usart3;
+	};
+
+	chosen {
+		stdout-path = &uart4;
+	};
+
+	led-controller-0 {
+		compatible = "gpio-leds";
+
+		led-0 {
+			label = "tac:green:user1";
+			gpios = <&gpiof 10 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
+		};
+
+		led-1 {
+			label = "tac:green:user2";
+			gpios = <&gpiog 7 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-2 {
+			label = "tac:green:statusdut";
+			gpios = <&gpioa 13 GPIO_ACTIVE_LOW>;
+		};
+
+		/* led-3 and led-4 are internally connected antiparallel to one
+		 * another inside the ethernet jack like this:
+		 * GPIOA14 ---+---|led-3|>--+--- GPIOD15
+		 *            +--<|led-4|---+
+		 * E.g. only one of the LEDs can be illuminated at a time while
+		 * the other output must be driven low.
+		 * This should likely be implemented using a multi color LED
+		 * driver for antiparallel LEDs. */
+		led-3 {
+			label = "tac:green:statuslab";
+			gpios = <&gpioa 14 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-4 {
+			label = "tac:orange:statuslab";
+			gpios = <&gpiod 15 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		user-button1 {
+			label = "USER_BTN";
+			linux,code = <KEY_HOME>;
+			gpios = <&gpioi 11 GPIO_ACTIVE_LOW>;
+		};
+
+		user-button2 {
+			label = "USER_BTN2";
+			linux,code = <KEY_ESC>;
+			gpios = <&gpioe 7 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	/* supplied by either barrel connector or PoE */
+	reg_12v: regulator_12v {
+		compatible = "regulator-fixed";
+		regulator-name = "12V";
+		regulator-min-microvolt = <12000000>;
+		regulator-max-microvolt = <12000000>;
+		regulator-always-on;
+	};
+
+	reg_5v: regulator_5v {
+		compatible = "regulator-fixed";
+		regulator-name = "5V";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-always-on;
+		vin-supply = <&reg_12v>;
+	};
+
+	reg_1v2: regulator_1v2 {
+		compatible = "regulator-fixed";
+		regulator-name = "1V2";
+		regulator-min-microvolt = <1200000>;
+		regulator-max-microvolt = <1200000>;
+		regulator-always-on;
+		vin-supply = <&reg_5v>;
+	};
+
+	reg_pb_5v: regulator_pb_5v {
+		compatible = "regulator-fixed";
+		regulator-name = "5V_POWERBOARD";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-always-on;
+		vin-supply = <&reg_5v>;
+	};
+
+	reg_pb_3v3: regulator_pb_3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "3V3_POWERBOARD";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+		vin-supply = <&reg_pb_5v>;
+	};
+
+	output_iobus_12v {
+		compatible = "regulator-output";
+		vout-supply = <&reg_iobus_12v>;
+	};
+
+	output_vuart {
+		compatible = "regulator-output";
+		vout-supply = <&v3v3_hdmi>;
+	};
+
+	iobus_curr: current-sense-amplifier-0 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 5>;
+		#io-channel-cells = <0>;
+		label = "iobus-curr";
+		sense-resistor-micro-ohms = <2200000000>; /* 2.2k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	iobus_volt: voltage-divider {
+		compatible = "voltage-divider";
+		io-channels = <&adc1 9>;
+		#io-channel-cells = <0>;
+		label = "iobus-volt";
+		output-ohms = <1000>;
+		full-ohms = <6100>;
+	};
+
+	iobus {
+		compatible = "iio-hwmon";
+		io-channels = <&iobus_volt>, <&iobus_curr>;
+	};
+
+	host_curr: current-sense-amplifier-1 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 13>;
+		#io-channel-cells = <0>;
+		label = "host-curr";
+		sense-resistor-micro-ohms = <1300000000>; /* 1.3k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_1_curr: current-sense-amplifier-2 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 15>;
+		#io-channel-cells = <0>;
+		label = "host-1-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_2_curr: current-sense-amplifier-3 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 0>; /* via ANA0 */
+		#io-channel-cells = <0>;
+		label = "host-2-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_3_curr: current-sense-amplifier-4 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 1>; /* via ANA1 */
+		#io-channel-cells = <0>;
+		label = "host-3-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	usb {
+		compatible = "iio-hwmon";
+		io-channels = <&host_curr>, <&host_1_curr>, <&host_2_curr>, <&host_3_curr>;
+	};
+};
+
+&pwr_regulators {
+	vdd-supply = <&vdd>;
+	vdd_3v3_usbfs-supply = <&vdd_usb>;
+};
+
+&hash1 {
+	status = "okay";
+};
+
+&crc1 {
+	status = "okay";
+};
+
+&cryp1 {
+	status = "okay";
+};
+
+&vrefbuf {
+	regulator-min-microvolt = <2500000>;
+	regulator-max-microvolt = <2500000>;
+	vdda-supply = <&vdda>;
+	status = "okay";
+};
+
+&adc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&adc1_ain_pins_a>;
+	vdd-supply = <&vdd>;
+	vdda-supply = <&vdda>;
+	vref-supply = <&vrefbuf>;
+	status = "okay";
+
+	adc1: adc@0 {
+		st,adc-channels = <0 1 2 5 9 10 13 15>;
+		st,min-sample-time-nsecs = <5000>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "okay";
+		channel@0 {
+			reg = <0>;
+			label = "HOST_2_CURR_FB";
+		};
+		channel@1 {
+			reg = <1>;
+			label = "HOST_3_CURR_FB";
+		};
+		channel@2 {
+			reg = <2>;
+			label = "OUT_0_FB";
+		};
+		channel@5 {
+			reg = <5>;
+			label = "IOBUS_CURR_FB";
+		};
+		channel@9 {
+			reg = <9>;
+			label = "IOBUS_VOLT_FB";
+		};
+		channel@10 {
+			reg = <10>;
+			label = "OUT_1_FB";
+		};
+		channel@13 {
+			reg = <13>;
+			label = "HOST_CURR_FB";
+		};
+		channel@15 {
+			reg = <15>;
+			label = "HOST_1_CURR_FB";
+		};
+	};
+
+	adc2: adc@100 {
+		st,adc-channels = <12>;
+		st,min-sample-time-nsecs = <500000>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "okay";
+		channel@12 {
+			reg = <12>;
+			label = "TEMP_INTERNAL";
+		};
+	};
+};
+
+&rtc {
+	status = "okay";
+};
+
+&iwdg2 {
+	timeout-sec = <8>;
+	status = "okay";
+};
+
+&timers2 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+	timer@1 {
+		status = "okay";
+	};
+};
+
+&timers3 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+	timer@2 {
+		status = "okay";
+	};
+};
+
+&timers4 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+	timer@3 {
+		status = "okay";
+	};
+};
+
+&dts {
+	status = "okay";
+};
+
+&gpiob {
+	gpio-line-names = "", "", "", "", "", "", "USB_RESET", "", \
+	"", "", "", "", "", "", "", "";
+};
+
+&gpiod {
+	gpio-line-names = "", "", "", "", "TP38", "TP39", "", "", \
+	"TP41", "TP42", "OLED_DC", "", "", "ETH_CS", "", "ETH_LAB_LEDRN";
+};
+
+&gpioe {
+	gpio-line-names = "TP35", "", "", "", "CAN_1_120R", "", "", "USER_BTN2", \
+	"TP48", "UART_TX_EN", "UART_RX_EN", "TP24", "", "TP25", "TP26", "TP27";
+};
+
+&gpiof {
+	gpio-line-names = "TP36", "TP37", "", "", "OLED_CS", "", "", "", \
+	"", "", "USER_LED1", "", "STACK_CS0", "", "", "";
+};
+
+&gpiog {
+	gpio-line-names = "ETH_RESET", "", "", "", \
+	"", "IOBUS_FLT_FB", "", "USER_LED2", \
+	"ETH1_PPS_A", "CAN_0_120R", "TP49", "", \
+	"", "", "", "";
+};
+
+&gpioh {
+	gpio-line-names = "", "", "OUT_1", "OUT_0", "OLED_RESET", "", "", "", \
+	"", "", "ETH1_PPS_B", "ETH_GPIO2", "", "IOBUS_PWR_EN", "", "TP33";
+};
+
+&gpioi {
+	gpio-line-names = "TIM_RTS", "", "", "", \
+	"DEVICE_DATA_EN", "", "", "", \
+	"ETH_WOL", "TP43", "", "USER_BTN";
+};
+
+&gpioz {
+	gpio-line-names = "HWID0", "HWID1", "HWID2", "HWID3", \
+	"", "", "HWID4", "HWID5";
+};
+
+&pmic {
+	regulators {
+		buck1-supply = <&reg_5v>;	/* VIN */
+		buck2-supply = <&reg_5v>;	/* VIN */
+		buck3-supply = <&reg_5v>;	/* VIN */
+		buck4-supply = <&reg_5v>;	/* VIN */
+		ldo2-supply = <&reg_5v>;	/* PMIC_LDO25IN */
+		ldo4-supply = <&reg_5v>;	/* VIN */
+		ldo5-supply = <&reg_5v>;	/* PMIC_LDO25IN */
+		vref_ddr-supply = <&reg_5v>;	/* VIN */
+		boost-supply = <&reg_5v>;	/* PMIC_BSTIN */
+		pwr_sw2-supply = <&bst_out>;	/* PMIC_SWIN */
+	};
+};
+
+&sdmmc2 {
+	pinctrl-names = "default", "opendrain", "sleep";
+	pinctrl-0 = <&sdmmc2_b4_pins_a &sdmmc2_d47_pins_b>;
+	pinctrl-1 = <&sdmmc2_b4_od_pins_a &sdmmc2_d47_pins_b>;
+	pinctrl-2 = <&sdmmc2_b4_sleep_pins_a &sdmmc2_d47_sleep_pins_b>;
+	bus-width = <8>;
+	mmc-ddr-3_3v;
+	no-1-8-v;
+	no-sd;
+	no-sdio;
+	non-removable;
+	st,neg-edge;
+	vmmc-supply = <&v3v3>;
+	status = "okay";
+};
+
+&uart4 {
+	label = "debug";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_pins_a>;
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	status = "okay";
+};
+
+&ethernet0 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&ethernet0_rgmii_pins_d>;
+	pinctrl-1 = <&ethernet0_rgmii_sleep_pins_d>;
+	phy-mode = "rgmii-id";
+	assigned-clocks = <&rcc ETHCK_K>, <&rcc PLL4_P>;
+	assigned-clock-parents = <&rcc PLL4_P>;
+	assigned-clock-rates = <125000000>; /* Clock PLL4 to 750Mhz in ATF/U-Boot */
+	st,eth-clk-sel;
+	status = "okay";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
+&i2c1 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&i2c1_pins_b>;
+	pinctrl-1 = <&i2c1_sleep_pins_b>;
+	status = "okay";
+
+	powerboard_eeprom: eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		vcc-supply = <&v3v3>;
+	};
+
+	lm75ad@48 {
+		compatible = "national,lm75a";
+		reg = <0x48>;
+		status = "disabled";
+	};
+};
+
+baseboard_eeprom: &sip_eeprom {
+};
+
+&i2c5 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&i2c5_pins_b>;
+	pinctrl-1 = <&i2c5_sleep_pins_b>;
+	status = "okay";
+	/delete-property/dmas;
+	/delete-property/dma-names;
+
+	usbhub: usbhub@2c {
+		compatible ="microchip,usb2514b";
+		reg = <0x2c>;
+		reset-gpios = <&gpiob 6 GPIO_ACTIVE_LOW>;
+		vdd-supply = <&v3v3>;
+	};
+};
+
+&spi2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi2_pins_b>;
+	cs-gpios = <&gpiof 12 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&spi5 {
+	/* spare dmas for other usage */
+	/delete-property/dmas;
+	/delete-property/dma-names;
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi5_pins_a>;
+	cs-gpios = <&gpiod 13 GPIO_ACTIVE_LOW>;
+	status = "okay";
+
+	ksz9563: ksz9563@0 {
+		compatible = "microchip,ksz9563";
+		reg = <0>;
+
+		reset-gpios = <&gpiog 0 GPIO_ACTIVE_LOW>;
+		spi-max-frequency = <44000000>;
+		spi-cpha;
+		spi-cpol;
+
+		interrupt-parent = <&gpioa>;
+		interrupts = <6 IRQ_TYPE_EDGE_RISING>;
+
+		status = "okay";
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			port_dut: port@0 {
+				reg = <0>;
+				label = "dut";
+			};
+			port_uplink: port@1 {
+				reg = <1>;
+				label = "uplink";
+			};
+			port_cpu: port@2 {
+				reg = <2>;
+				label = "cpu";
+
+				ethernet = <&ethernet0>;
+
+				phy-mode = "rgmii-id";
+				rx-internal-delay-ps = <2000>;
+				tx-internal-delay-ps = <2000>;
+
+				fixed-link {
+					speed = <1000>;
+					full-duplex;
+				};
+			};
+		};
+	};
+};
+
+&usbotg_hs {
+	phys = <&usbphyc_port1 0>;
+	phy-names = "usb2-phy";
+	dr_mode = "peripheral";
+	vusb_d-supply = <&vdd_usb>;
+	vusb_a-supply = <&reg18>;
+	status = "okay";
+};
+
+&usbh_ehci {
+	phys = <&usbphyc_port0>;
+	phy-names = "usb";
+	status = "okay";
+};
+
+&spi4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi4_pins_a>;
+	cs-gpios = <&gpiof 4 GPIO_ACTIVE_LOW>;
+	status = "okay";
+
+	lcd: display@0 {
+		compatible = "shineworld,lh133k", "panel-mipi-dbi-spi";
+		spi-max-frequency = <32000000>;
+		power-supply = <&v3v3>;
+		io-supply = <&v3v3>;
+		backlight = <&backlight>;
+
+		reg = <0>;
+		width-mm = <23>;
+		height-mm = <23>;
+		rotation = <180>;
+
+		reset-gpios = <&gpioh 4 GPIO_ACTIVE_HIGH>;
+		dc-gpios = <&gpiod 10 GPIO_ACTIVE_HIGH>;
+		spi-3wire;
+
+		panel-timing {
+			hactive = <240>;
+			vactive = <240>;
+			hback-porch = <0>;
+			vback-porch = <0>;
+
+			clock-frequency = <0>;
+			hfront-porch = <0>;
+			hsync-len = <0>;
+			vfront-porch = <0>;
+			vsync-len = <0>;
+		};
+	};
+};
+
+&usbphyc {
+	status = "okay";
+};
+
+&usbphyc_port0 {
+	phy-supply = <&vdd_usb>;
+};
+
+&usbphyc_port1 {
+	phy-supply = <&vdd_usb>;
+};
+
+&m_can1 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&m_can1_pins_b>;
+	pinctrl-1 = <&m_can1_sleep_pins_b>;
+	status = "okay";
+};
+
+&m_can2 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&m_can2_pins_a>;
+	pinctrl-1 = <&m_can2_sleep_pins_a>;
+	status = "okay";
+};
+
+&usart3 {
+	label = "dut";
+	uart-has-rtscts;
+
+	/delete-property/dmas;
+	/delete-property/dma-names;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&usart3_pins_e>;
+
+	status = "okay";
+};
+
+&pinctrl {
+	adc1_ain_pins_a: adc1-ain-0 {
+		pins {
+			pinmux = <STM32_PINMUX('F', 11, ANALOG)>, /* ADC1_INP2 */
+				 <STM32_PINMUX('B', 1, ANALOG)>, /* ADC1_INP5 */
+				 <STM32_PINMUX('B', 0, ANALOG)>, /* ADC1_INP9 */
+				 <STM32_PINMUX('C', 0, ANALOG)>, /* ADC1_INP10 */
+				 <STM32_PINMUX('C', 3, ANALOG)>, /* ADC1_INP13 */
+				 <STM32_PINMUX('A', 3, ANALOG)>; /* ADC1_INP15 */
+		};
+	};
+	ethernet0_rgmii_pins_d: rgmii-1 {
+		pins1 {
+			pinmux = <STM32_PINMUX('G', 4, AF11)>, /* ETH_RGMII_GTX_CLK */
+				 <STM32_PINMUX('G', 13, AF11)>, /* ETH_RGMII_TXD0 */
+				 <STM32_PINMUX('G', 14, AF11)>, /* ETH_RGMII_TXD1 */
+				 <STM32_PINMUX('C', 2, AF11)>, /* ETH_RGMII_TXD2 */
+				 <STM32_PINMUX('E', 2, AF11)>, /* ETH_RGMII_TXD3 */
+				 <STM32_PINMUX('B', 11, AF11)>; /* ETH_RGMII_TX_CTL */
+			bias-disable;
+			drive-push-pull;
+			slew-rate = <1>;
+		};
+		pins2 {
+			pinmux = <STM32_PINMUX('C', 4, AF11)>, /* ETH_RGMII_RXD0 */
+				 <STM32_PINMUX('C', 5, AF11)>, /* ETH_RGMII_RXD1 */
+				 <STM32_PINMUX('H', 6, AF11)>, /* ETH_RGMII_RXD2 */
+				 <STM32_PINMUX('H', 7, AF11)>, /* ETH_RGMII_RXD3 */
+				 <STM32_PINMUX('A', 1, AF11)>, /* ETH_RGMII_RX_CLK */
+				 <STM32_PINMUX('A', 7, AF11)>; /* ETH_RGMII_RX_CTL */
+			bias-disable;
+		};
+	};
+	ethernet0_rgmii_sleep_pins_d: rgmii-sleep-1 {
+		pins1 {
+			pinmux = <STM32_PINMUX('G', 4, ANALOG)>, /* ETH_RGMII_GTX_CLK */
+				 <STM32_PINMUX('G', 13, ANALOG)>, /* ETH_RGMII_TXD0 */
+				 <STM32_PINMUX('G', 14, ANALOG)>, /* ETH_RGMII_TXD1 */
+				 <STM32_PINMUX('C', 2, ANALOG)>, /* ETH_RGMII_TXD2 */
+				 <STM32_PINMUX('E', 2, ANALOG)>, /* ETH_RGMII_TXD3 */
+				 <STM32_PINMUX('B', 11, ANALOG)>, /* ETH_RGMII_TX_CTL */
+				 <STM32_PINMUX('C', 4, ANALOG)>, /* ETH_RGMII_RXD0 */
+				 <STM32_PINMUX('C', 5, ANALOG)>, /* ETH_RGMII_RXD1 */
+				 <STM32_PINMUX('H', 6, ANALOG)>, /* ETH_RGMII_RXD2 */
+				 <STM32_PINMUX('H', 7, ANALOG)>, /* ETH_RGMII_RXD3 */
+				 <STM32_PINMUX('A', 1, ANALOG)>, /* ETH_RGMII_RX_CLK */
+				 <STM32_PINMUX('A', 7, ANALOG)>; /* ETH_RGMII_RX_CTL */
+		 };
+	};
+	spi2_pins_b: spi2-1 {
+		pins1 {
+			pinmux = <STM32_PINMUX('I', 1, AF5)>, /* SPI2_SCK */
+				 <STM32_PINMUX('I', 3, AF5)>; /* SPI2_MOSI */
+			bias-disable;
+			drive-push-pull;
+		};
+
+		pins2 {
+			pinmux = <STM32_PINMUX('I', 2, AF5)>; /* SPI2_MISO */
+			bias-pull-down;
+		};
+	};
+	spi5_pins_a: spi5-0 {
+		pins1 {
+			pinmux = <STM32_PINMUX('F', 7, AF5)>, /* SPI5_SCK */
+				 <STM32_PINMUX('F', 9, AF5)>; /* SPI5_MOSI */
+			bias-disable;
+			drive-push-pull;
+			slew-rate = <1>;
+		};
+
+		pins2 {
+			pinmux = <STM32_PINMUX('F', 8, AF5)>; /* SPI5_MISO */
+			bias-disable;
+		};
+	};
+	usart3_pins_e: usart3-4 {
+		pins1 {
+			pinmux = <STM32_PINMUX('B', 10, AF7)>, /* USART3_TX */
+				 <STM32_PINMUX('D', 12, AF7)>; /* USART3_RTS */
+			bias-disable;
+			drive-push-pull;
+			slew-rate = <0>;
+		};
+		pins2 {
+			pinmux = <STM32_PINMUX('B', 12, AF8)>, /* USART3_RX */
+				 <STM32_PINMUX('D', 11, AF7)>; /* USART3_CTS_NSS */
+			bias-disable;
+		};
+	};
+};
